### PR TITLE
[inductor][cpp] support vectorized indirect indexing

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119979
* #119734
* __->__ #119655
* #119654

This PR adds the vectorized indirect indexing so that we can further simplify the `CppVecKernelChecker` (done in the later PR #119734) and remove the check that throws `CppVecUnsupportedError`. A boundary assertion check is added on vectorized indices and via the new `indirect_assert` method on `Kernel` - the base implementation is for scalar indices, overridden in `CppVecKernel` for vectorized indices.

cc @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/utils.py [hunk 1]
```diff
@@ -874,15 +874,21 @@ def do_indent(self, offset=1):
     def do_unindent(self, offset=1):
         self._indent -= offset
 
-    def splice(self, other_code, strip=False):
+    def splice(self, other_code, strip=False, line_wrapper=None):
```
Comment (jgong5) [inline torch/_inductor/utils.py:877]: This new `line_wrapper` allows the `splice` on deferred lines.

Comment (jansel) [reply torch/_inductor/utils.py:877]: Hrm, the semantics of `line_wrapper` aren't super clear to me.  Could we make `splice` jut support DeferredLine directly?

If we need to wrap every line in an IndentedBuffer, perhaps we could add a seperate `.wrap()` method?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/utils.py (hunk 1 above)
Comment (jgong5) [reply torch/_inductor/utils.py]: Thanks. Fixed.

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -874,15 +874,24 @@ def do_indent(self, offset=1):
     def do_unindent(self, offset=1):
         self._indent -= offset
 
-    def splice(self, other_code, strip=False):
+    def splice(self, other_code, strip=False, deferred_line_name=None):
+        def maybe_deferred_line(line):
+            if deferred_line_name is not None and not isinstance(line, LineContext):
+                from torch._inductor.codegen.common import DeferredLine
+
+                return DeferredLine(deferred_line_name, line)
```
Comment (jansel) [inline torch/_inductor/utils.py:882]: I don't think `.splice()` should have the side effect of wrapping everything being spliced in a `DeferredLine`.  I was suggesting before doing it in two steps, first wrap a code fragment with DeferredLine (this code fragmenet can be a seperate IndentedBuffer).  Next splice the other IndentedBuffer into the larger one.

**Review by jansel — CHANGES_REQUESTED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**